modules/transactions/src/main/java/org/apache/ignite/internal/tx/storage/state/rocksdb/TxStateRocksDbStorage.java

```diff
@@ -392,64 +259,225 @@ public long persistedIndex() {
+    private class StorageIterator implements Iterator<IgniteBiTuple<UUID, TxMeta>>, AutoCloseable {
+        /** Busy lock. */
+        private final IgniteSpinBusyLock busyLock;
+
+        /** Rocks iterator. */
+        private final RocksIterator rocksIterator;
+
+        /** Current value. */
+        private IgniteBiTuple<UUID, TxMeta> cur;
+
+        /** Whether this iterator is valid. It becomes invalid when reaches data that belongs to abother partition. */
+        private boolean isValid = true;
+
+        /**
+         * Constructor.
+         *
+         * @param busyLock Busy lock.
+         * @param rocksIterator RocksDB iterator.
+         */
+        private StorageIterator(IgniteSpinBusyLock busyLock, RocksIterator rocksIterator) {
+            this.busyLock = busyLock;
+            this.rocksIterator = rocksIterator;
+        }
+
+        /** {@inheritDoc} */
+        @Override
+        public boolean hasNext() {
+            if (!busyLock.enterBusy()) {
+                throwStorageStoppedException();
+            }
+
+            try {
+                if (!isValid) {
+                    return false;
+                }
+
+                advance();
+
+                return cur != null;
+            } finally {
+                busyLock.leaveBusy();
+            }
+        }
+
+        /** {@inheritDoc} */
+        @Override
+        public IgniteBiTuple<UUID, TxMeta> next() {
+            if (!busyLock.enterBusy()) {
+                throwStorageStoppedException();
+            }
+
+            try {
+                if (!isValid) {
+                    throw new NoSuchElementException();
+                }
+
+                advance();
+
+                if (cur != null) {
+                    IgniteBiTuple<UUID, TxMeta> temp = cur;
+
+                    cur = null;
+
+                    return temp;
+                } else {
+                    throw new NoSuchElementException();
+                }
+            } finally {
+                busyLock.leaveBusy();
+            }
+        }
+
+        private void advance() {
+            if (cur != null) {
+                return;
+            }
+
+            boolean isRocksIteratorValid = rocksIterator.isValid();
+
+            if (!isRocksIteratorValid) {
+                checkIterator(rocksIterator);
+
+                return;
+            }
+
+            while (true) {
+                rocksIterator.next();
+
+                byte[] keyBytes = rocksIterator.key();
+                byte[] valueBytes = rocksIterator.value();
+
+                if (keyBytes.length == Integer.BYTES) {
+                    int p = ByteBuffer.wrap(keyBytes).getInt();
+
+                    if (p == partitionId) {
+                        // Skip this entry, it's last applied index.
+                        continue;
+                    } else {
+                        // The data range is over, this is last applied index of the next partition.
+                        isValid = false;
```

Comment:
This part can be implemented with upper bound in iterator. You would save some code and make it simpler